oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreIterator.java

```diff
@@ -19,34 +19,44 @@
 
 package org.apache.jackrabbit.oak.index.indexer.document.flatfile;
 
-import static org.apache.jackrabbit.guava.common.collect.Iterators.concat;
-import static org.apache.jackrabbit.guava.common.collect.Iterators.singletonIterator;
-
-import java.io.Closeable;
-import java.util.Iterator;
-import java.util.Set;
-
+import org.apache.jackrabbit.guava.common.collect.AbstractIterator;
+import org.apache.jackrabbit.oak.commons.IOUtils;
 import org.apache.jackrabbit.oak.index.indexer.document.NodeStateEntry;
 import org.apache.jackrabbit.oak.index.indexer.document.flatfile.linkedList.FlatFileBufferLinkedList;
 import org.apache.jackrabbit.oak.index.indexer.document.flatfile.linkedList.NodeStateEntryList;
 import org.apache.jackrabbit.oak.index.indexer.document.flatfile.linkedList.PersistedLinkedList;
+import org.apache.jackrabbit.oak.index.indexer.document.flatfile.linkedList.PersistedLinkedListV2;
 import org.apache.jackrabbit.oak.index.indexer.document.flatfile.pipelined.ConfigHelper;
 import org.apache.jackrabbit.oak.spi.blob.BlobStore;
 import org.apache.jackrabbit.oak.spi.state.NodeState;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import org.apache.jackrabbit.guava.common.collect.AbstractIterator;
+import java.io.Closeable;
+import java.util.Iterator;
+import java.util.Set;
+
+import static org.apache.jackrabbit.guava.common.collect.Iterators.concat;
+import static org.apache.jackrabbit.guava.common.collect.Iterators.singletonIterator;
 
 class FlatFileStoreIterator extends AbstractIterator<NodeStateEntry> implements Iterator<NodeStateEntry>, Closeable {
-    private static final Logger log = LoggerFactory.getLogger(FlatFileStoreIterator.class);
+    private static final Logger LOG = LoggerFactory.getLogger(FlatFileStoreIterator.class);
 
     static final String BUFFER_MEM_LIMIT_CONFIG_NAME = "oak.indexer.memLimitInMB";
     // by default, use the PersistedLinkedList
     private static final int DEFAULT_BUFFER_MEM_LIMIT_IN_MB = 0;
-    static final String PERSISTED_LINKED_LIST_CACHE_SIZE = "oak.indexer.persistedLinkedList.cacheSize";
-    static final int DEFAULT_PERSISTED_LINKED_LIST_CACHE_SIZE = 1000;
 
+    public static final String PERSISTED_LINKED_LIST_CACHE_SIZE = "oak.indexer.persistedLinkedList.cacheSize";
+    public static final int DEFAULT_PERSISTED_LINKED_LIST_CACHE_SIZE = 1000;
+
+    public static final String PERSISTED_LINKED_LIST_V2_CACHE_SIZE = "oak.indexer.persistedLinkedListV2.cacheSize";
+    public static final int DEFAULT_PERSISTED_LINKED_LIST_V2_CACHE_SIZE = 10000;
+
+    public static final String PERSISTED_LINKED_LIST_V2_MEMORY_CACHE_SIZE_MB = "oak.indexer.persistedLinkedListV2.cacheMaxSizeMB";
+    public static final int DEFAULT_PERSISTED_LINKED_LIST_V2_MEMORY_CACHE_SIZE_MB = 8;
+
+    public static final String PERSISTED_LINKED_LIST_USE_V2 = "oak.indexer.persistedLinkedList.useV2";
```

Comment:
Since we are using versioning convention here - would it make more sense to actually configure the version that is to be used instead of a flag to use/not use a particular version ?

Reply:
I don't expect that there will be more versions of PersistedLinkedList. My intention is that for some time we keep both versions around, with the possibility to chose one or the other via configuration. It is a safe-guard for the case we find some issue with the new implementation. But once we feel confident that the PersistedLinkedListV2 is working properly, we should remove the old version as it offers no advantage over the new one, and also remove this configuration property.

Reply:
(As always when using caches with a "size" limit and not a "memory" limit) there was a risk of running out-of-memory. This risk existed even before, with 1000. 

Now we have 10'000, but there is another "memory" limit at 8 MB. 8 MB is s bit small in my view, I would use 32 MB at least. What about using just the memory limit, and not a size limit? That would simplify the code, and simpler code typically has less bugs. I understand that on the other hand, if there is a bug with the memory calculation, then there is a risk as well. But I think the memory calculation must not have bugs (of the form, memory is smaller or equal 0): if it has, then we are in trouble anyway.

So I would opt for the simpler code!

Reply:
Both of these design choices are based on learnings from developing the Pipelined strategy, which internally also keeps several buffers that ideally should be as big as possible while remaining within the limits.

- 8MB default - This is intentionally low to avoid failures when running with small heaps. In the past, I had test failures in some CI agents because they were configured with only 512MB of RAM and I had set a default buffer size that was taking up a significant portion of the available memory. The intention is that in any deployment of Oak, the value will be set to 128MB or greater, but by default, I prefer to be very conservative. And 8MB is not so small, based on my experiments, 1000 entries usually take less than 1MB of RAM and this is the default for the previous implementation. 
- Limit on size and count - This is to account for the case where entries are all very small. The memory estimation is not perfect, because it does not take in consideration the memory required by the internal structures of the HashMap or any other per-entry overhead that we may not be considering. Once again, I had issues with unexpected OOME with the buffers used in the Pipelined strategy when I did not set a limit on the number of entries. And having a limit on the number of entries gives us another lever to adjust in case we find some unexpected issues in production.

Reply:
> This is to account for the case where entries are all very small.

What I usually do is add some offset to the memory calculation... basically "+ 100" or so. (in the past I also verified that the memory calculation is somewhat correct, by have a small program that runs out-of-memory... but I understand this can't be a unit test, and we want to support multiple JVM versions without having to test each time... so the memory estimation of course won't ever be accurate... but it doesn't have to be)

Reply:
The memory estimation is conservative because it assumes that all strings will be encoded with 2 bytes per char. In practice, in most cases the majority of strings will be encoded with 1 byte per char, so the memory estimation will almost double than the memory effectively used. And if using string interning or string deduplication, the difference might be even bigger. But of course we have to do the calculation for the worst case scenario. And as you say, it does not need to be precise. The goal here is just to prevent OOME while ensuring that the cache is big enough to give us the majority of the performance benefits.